test/e2e/undo-input.spec.ts

```diff
@@ -0,0 +1,55 @@
+import { expect } from '@playwright/test';
+import { test } from './electron-test';
+import {saveLastBlockAndEnter, waitForBoot, createPage, deleteCurrentPage, isMac} from "./utils";
+
+const modifier = isMac ? 'Meta' : 'Control';
+const undoShortcut = `${modifier}+Z`;
+const redoShortcut = `${modifier}+Shift+Z`;
+
+test('undo input test', async ({ page }) => {
+    // Setup.
+    await waitForBoot(page);
+    await createPage(page, 'undo input');
+
+    // Make three blocks, but don't make a fourth.
+    // Wait a bit for transaction and focus events to be resolved between input.
+    await saveLastBlockAndEnter(page, "one");
+    await saveLastBlockAndEnter(page, "two");
+    await saveLastBlockAndEnter(page, "three");
+
+    // All three blocks should be there.
+    await expect(page.locator('.block:has-text("one")')).toBeVisible();
+    await expect(page.locator('.block:has-text("two")')).toBeVisible();
+    await expect(page.locator('.block:has-text("three")')).toBeVisible();
+
+    // Undo once, removing the last block input and split.
+    await page.keyboard.press(undoShortcut);
+
+    // Last block should be gone.
+    await expect(page.locator('.block:has-text("one")')).toBeVisible();
```

Comment:
How is `expect` async here? I thought this would have to be written as

`expect(await page.locator('.block:has-text("one")')).toBeVisible();`

Reply:
I think the expectation itself is async, because I know it can wait a certain amount of time for it to be or not be visible. If the locator itself was the async part here, I don't think it could wait to not be visible.